scripts/backup/run_backup/README_IMPLEMENTATION.txt

```diff
@@ -0,0 +1,67 @@
+The backup script is implemented in Python. The following extra
+modules are needed:
+
+psycopg2    - PostgresQL driver 
+boto3       - AWS sdk, for accessing S3 storage
+paramiko    - SSH client, for transferring files via SFTP
+swiftclient - for reading [not yet implemented] and writing swift objects. 
+
+1. Database access.
+
+The module uses psycopg2 to access the Dataverse database, to obtain
+the lists of files that have changed since the last backup that need
+to be copied over. Additionally, it maintains its own database for
+keeping track of the backup status of individual files. As of now,
+this extra database must reside on the same server as the main
+Dataverse database and is owned by the same Postgres user.
+
+Consult README_HOWTO.txt on how to set up this backup database (needs
+to be done prior to running the backup script)
+
+2. Storage access
+
+Currently implemented storage access methods, for local filesystem and
+S3 are isolated in the files storage_filesystem.py and storage_s3.py,
+respectively. To add support for swift a similar fragment of code will
+need to be provided, with an open_storage_object... method that can go
+to the configured swift end node and return the byte stream associated
+with the datafile. Use storage_filesystem.py as the model. Then the
+top-level storage.py class will need to be modified to import and use
+the extra storage method.
+
+3. Backup (write) acces.
+
+Similarly, storage type-specific code for writing backed up objects is
+isolated in the backup_...py files. The currently implemented storage
+methods are ssh/ftp (backup_ssh.py, default) and swift
+(backup_swift.py; experimental, untested). To add support for other
+storage systems, use backup_ssh.py as the model to create your own
+backup_... classes, implementing similar methods, that a) copy the
+byte stream associated with a Dataverse datafile onto this storage
+system and b) verify the copy against the checksum (MD5 or SHA1)
+provided by the Dataverse.  In the SSH/SFTP implementation, we can do
+the verification step by simply executing md5sum/sha1sum on the remote
+server via ssh, once the file is copied. With swift, the only way to
+verify against the checksum is to read the file *back* from the swift
+end note, and calculate the checksum on the obtained stream.
```

Comment:
I'm glad that there is some discussion of and support for checksums of the backed up data to help assure data integrity. The comments about Swift make me wonder if Harvard Dataverse or any installation will go the extra mile to read back the files from Swift for the purpose of verifying the data integrity.